Add Enterprise Search API endpoints for 1 Click ELSER ML Model Deployment

markjhoy opened the pull request:
## Summary

Adds Enterprise Search internal API endpoints for deploying and monitoring the deployment status of an ELSER ML model (and possibly other models in the future) via the 1 click deployment process. This is to not allow a direct call from the Kibana front end to the underlying Elasticsearch ML endpoints.

Closes https://github.com/elastic/enterprise-search-team/issues/4295 and https://github.com/elastic/enterprise-search-team/issues/4397

### Checklist
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

demjened commented:
@markjhoy Great first draft, thanks!

There are a couple of issues we need to sort out before ELSER download process can work properly:
- In `get_ml_model_deployment_status.ts` the `trainedModelsProvider.getTrainedModels()` call returns a 404 error if the model does not exist, so we need to wrap it in a try/catch block rather than checking `modelDetailsResponse`. We should  check for `error.statusCode !== 404` within the catch block; this is what that API returns upon a missing model.
- While the model is downloading (but not deployed), the `modelDetailsResponse.trained_model_configs` property does not exist. The correct check is `!modelDetailsResponse.trained_model_configs || !modelDetailsResponse.trained_model_configs[0].fully_defined`.
- We'll need to make sure the types in the ML JS client have been updated before this PR can be merged: `MlTrainedModelConfig.fully_defined`, `MlInclude.definition_status`.

markjhoy commented:
> * In `get_ml_model_deployment_status.ts` the `trainedModelsProvider.getTrainedModels()` call returns a 404 error if the model does not exist, so we need to wrap it in a try/catch block rather than checking `modelDetailsResponse`. We should  check for `error.statusCode !== 404` within the catch block; this is what that API returns upon a missing model.

Ah - gotcha - I wasn't 100% certain.... updated to check this.

> * While the model is downloading (but not deployed), the `modelDetailsResponse.trained_model_configs` property does not exist. The correct check is `!modelDetailsResponse.trained_model_configs || !modelDetailsResponse.trained_model_configs[0].fully_defined`.

Good catch. Updated.

> * We'll need to make sure the types in the ML JS client have been updated before this PR can be merged: `MlTrainedModelConfig.fully_defined`, `MlInclude.definition_status`.

Yep - hopefully that'll come soon...

> * Are you sure the new deployment state is called `is_fully_downloaded`, not `fully_downloaded`? The latter is more in line with the other states, e.g. `fully_allocated`.

I think that was some of the initial wording that was there, but subsequently changed... I've cleaned up the code better to reflect the current state of things. 👍

markjhoy commented on x-pack/plugins/enterprise_search/server/lib/ml/start_ml_model_download.ts:
```diff
@@ -0,0 +1,90 @@
+  await mlClient.putTrainedModel(putRequest);
+
+  // and sync our objects
+  await syncSavedObjectsFactory(clusterClient, savedObjectService).syncSavedObjects(false);
```
Oh, nice. Thanks for the information. So - when we do the `putTrainedModel` call here, it will automatically sync for us then?

sphilipse commented on x-pack/plugins/enterprise_search/server/lib/ml/start_ml_model_deployment.ts:
```diff
@@ -0,0 +1,74 @@
+    throw new Error('Machine Learning is not enabled');
+  }
+
+  let deploymentState: MlModelDeploymentState = MlModelDeploymentState.NotDeployed;
```
nitpick: you can just remove this initial assigment, it's only assigned once below.

markjhoy replied:
Good catch - this was from when the code was doing a lot more before a couple of commits ago...

sphilipse commented on x-pack/plugins/enterprise_search/server/routes/enterprise_search/indices.ts:
```diff
@@ -989,4 +994,165 @@ export function registerIndexRoutes({
+        true,
+        undefined,
+        client,
+        () => Promise.resolve() // pretend isMlReady, to allow us to initialize the saved objects
```
question: this feels a bit icky, what if ml isn't ready?

markjhoy replied:
This was actually lifted from some ML-side code itself... but, if we are able to call this function, the ML module _will_ be guaranteed to be ready, so I think it's OK to leave this here (although I'll remove the comment to lessen confusion ;) )

sphilipse commented on x-pack/plugins/enterprise_search/server/routes/enterprise_search/indices.ts:
```diff
@@ -989,4 +994,165 @@ export function registerIndexRoutes({
+        ? await ml.trainedModelsProvider(request, savedObjectsClient)
+        : undefined;
+
+      const savedObjectService = mlSavedObjectServiceFactory(
```
We're doing the same (or very similar) initialization and wrapping to check a 404 error in all three endpoints. Can we generalize this to a separate wrapping function?

markjhoy replied:
good call on these... I didn't wrap things, but I did clean up and have a common function to check this

sphilipse reviewed (commented):
Cursory review with a few nitpicks and question, but looks good overall.

markjhoy commented on x-pack/plugins/enterprise_search/server/lib/ml/start_ml_model_download.ts:
```diff
@@ -0,0 +1,81 @@
+    throw new Error('Machine Learning is not enabled');
+  }
+
+  let deploymentState: MlModelDeploymentState = MlModelDeploymentState.NotDeployed;
```
👍  same as above

markjhoy commented:
@elasticmachine merge upstream

markjhoy commented on x-pack/plugins/enterprise_search/server/routes/enterprise_search/indices.ts:
```diff
@@ -989,4 +994,159 @@ export function registerIndexRoutes({
+        () => Promise.resolve()
+      );
+
+      const mlClient = savedObjectService
```
Ah --- I didn't see the `ml.` object earlier (and some of this code was left over from when we thought we needed to sync/save the objects too)... thanks for the heads up, I'll correct this (which will also remove the need for including the `getMlClient` function below as well!)

jgowdyelastic replied:
The `trainedModelsProvider` is currently missing some types which you need, I have a [PR](https://github.com/elastic/kibana/pull/155822) waiting to merge which adds them in, so you'll need to update this PR to be able to use them

jgowdyelastic commented on x-pack/plugins/enterprise_search/server/lib/ml/start_ml_model_download.ts:
```diff
@@ -0,0 +1,69 @@
+  };
+
+  // this will also sync our saved objects for us
+  await esClient.ml.putTrainedModel(putRequest);
```
ml endpoints can't be called directly from elasticsearch. you need to use the functions shared in `trainedModelsProvider` 
e.g. `trainedModelsProvider.putTrainedModel`

There is currently a [PR](https://github.com/elastic/kibana/pull/155822) open to correct the types, once this is merged you'll be able see `putTrainedModel`. but if you need to test before that, `putTrainedModel` does exist, it's just that the types say it doesn't.

demjened reviewed (approved):
LGTM

TattdCodeMonkey commented on x-pack/plugins/enterprise_search/server/lib/ml/get_ml_model_deployment_status.ts:
```diff
@@ -0,0 +1,119 @@
+  trainedModelsProvider: MlTrainedModels | undefined
+): Promise<MlModelDeploymentStatus> => {
+  if (!trainedModelsProvider) {
+    throw new Error('Machine Learning is not enabled');
```
@demjened I think this is ok since we don't fetch model status if ml is disabled, but we should ensure thats the case.

TattdCodeMonkey reviewed (approved):
Overall LGTM, the TODOs and missing types make a little uneasy given where we are in release cycle.

markjhoy commented:
> the TODOs and missing types make a little uneasy given where we are in release cycle

100% agree... hopefully the missing types will be there before BC2 and we can alleviate these....

jgowdyelastic commented:
> > the TODOs and missing types make a little uneasy given where we are in release cycle
> 
> 100% agree... hopefully the missing types will be there before BC2 and we can alleviate these....

Just to confirm, there were two sets of missing types in this PR. The ones in the elasticsearch-js client, and the ones for ML's `TrainedModelsProvider`.
The `TrainedModelsProvider` types are now up to date and have been used in this PR.

The elasticsearch-js client types I do not believe will make this version as historically a client update comes with a lot of types churn throughout kibana.
These missing types are minimal, one missing param in the request and one property in the response of the get trained models api.

markjhoy commented:
> Just to confirm, there were two sets of missing types affecting this PR. The ones in the elasticsearch-js client, and the ones for ML's `TrainedModelsProvider`. The `TrainedModelsProvider` types are now up to date and have been used in this PR.
> 
> The elasticsearch-js client types I do not believe will make this version as historically a client update comes with a lot of types churn throughout kibana. These missing types are minimal, one missing param in the request and one property in the response of the get trained models api.

Thanks @jgowdyelastic  -- we also have a follow on task to check and update this in the future when things are available: https://github.com/elastic/enterprise-search-team/issues/4432

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create automatically backports add the label `auto-backport` or prevent reminders by adding the `backport:skip` label.
You can also create backports manually by running `node scripts/backport --pr 155213` locally

sphilipse commented:
Removing version/backport labels because it's too late to add backports now.

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create automatically backports add the label `auto-backport` or prevent reminders by adding the `backport:skip` label.
You can also create backports manually by running `node scripts/backport --pr 155213` locally

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create automatically backports add the label `auto-backport` or prevent reminders by adding the `backport:skip` label.
You can also create backports manually by running `node scripts/backport --pr 155213` locally
